### PR TITLE
For #10248 - Obfuscates credit card numbers autofill prompt

### PR DESCRIPTION
**Pull request by codrut-topliceanu**
Fixes https://github.com/mozilla-mobile/android-components/issues/10248

**Needs to land before https://github.com/mozilla-mobile/fenix/pull/19420**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks .
- [x] **Tests**: This PR includes thorough tests.
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md).

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/prompts/src/main/java/mozilla/components/feature/prompts/creditcard/CreditCardItemViewHolder.kt [hunk 1]
```diff
@@ -59,5 +62,10 @@ class CreditCardItemViewHolder(
 
         // Date format pattern for the credit card expiry date.
         private const val DATE_PATTERN = "MM/yyyy"
+
+        /**
+         * Number of digits to be displayed after ellipses on an obfuscated credit card number.
+         */
```
Comment (gabrielluong) [inline components/feature/prompts/src/main/java/mozilla/components/feature/prompts/creditcard/CreditCardItemViewHolder.kt:68]:
```suggestion
        // Number of digits to be displayed after ellipses on an obfuscated credit card number.
```

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt [hunk 2]
```diff
@@ -190,6 +190,26 @@ fun String.getDataUrlImageExtension(defaultExtension: String = "jpg"): String {
         .find(this)?.groups?.get(1)?.value ?: defaultExtension
 }
 
+/**
+ * Returns the input string preceded by 4 ellipses, locked LTR, to be used for
+ * obfuscating credit card numbers.
+ */
+fun String.addEllipsesToCreditCardNumber(): String {
```
Comment (gabrielluong) [inline components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt:197]: I thought our intention was to add a getter in the `CreditCard` data class to return the obfuscated card number? This seems ok, but maybe too specific for CreditCards. I will check in with some AC folks to see if there are any strong opinions.

Comment (Amejia481) [reply components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt:197]: I agree with @gabrielluong,  this function is too spesific to `CreditCard`  to be as `String` extension function. If we don't want to add the function directly to `CreditCard` we would make this to be an extension function of `CreditCard`.

Comment (gabrielluong) [reply components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt:197]: Will request changes then.

**Review by gabrielluong — CHANGES_REQUESTED** [review]

File: components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt (hunk 2 above)
Comment (codrut-topliceanu) [reply components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt:197]: @gabrielluong @Amejia481 I agree. I do want to add a getter, but there is more than one `crediCard` type object, as you mentioned, and I did not want to have the same code in two places. Should I just add the custom getter to both?

Comment (gabrielluong) [reply components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlin/String.kt:197]: I think either option is fine. The `CreditCard` in concept-engine is more specific to the GV Autocomplete `CreditCard`, which will have the credit card decrypted at that point. You will also need to get the last 4 digits for that one. So, it seems easier for me to imagine 2 different getters due to the additional need to get the last 4 digits for the concept-engine `CreditCard`. I don't really know what a custom getter for both will look like, so feel free to post both version and we can work through it together.

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/CreditCard.kt [hunk 3]
```diff
@@ -25,4 +25,25 @@ data class CreditCard(
     val expiryMonth: String,
     val expiryYear: String,
     val cardType: String
-) : Parcelable
+) : Parcelable {
+    val cardNumberLast4: String
+        get() = number.substring(number.length - digitsToShow)
+    val ellipsizedNumber: String
```
Comment (gabrielluong) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/CreditCard.kt:31]: Let's call this `obfuscatedCardNumber`

**Review by gabrielluong — CHANGES_REQUESTED** [review]

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/CreditCard.kt [hunk 4]
```diff
@@ -25,4 +25,25 @@ data class CreditCard(
     val expiryMonth: String,
     val expiryYear: String,
     val cardType: String
-) : Parcelable
+) : Parcelable {
+    val cardNumberLast4: String
```
[the remark this answers is not in the document]
Comment (codrut-topliceanu) [reply components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/CreditCard.kt:29]: >The CreditCard in concept-engine is more specific to the GV Autocomplete CreditCard, which will have the credit card decrypted at that point. You will also need to get the last 4 digits for that one.

-this is the only reason I put that there :P

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/CreditCard.kt [hunk 5]
```diff
@@ -25,4 +25,24 @@ data class CreditCard(
     val expiryMonth: String,
     val expiryYear: String,
     val cardType: String
-) : Parcelable
+) : Parcelable {
+    val obfuscatedCardNumber: String
```
Comment (gabrielluong) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/prompt/CreditCard.kt:29]: We have a test for this https://searchfox.org/mozilla-mobile/source/android-components/components/concept/engine/src/test/java/mozilla/components/concept/engine/prompt/CreditCardTest.kt#13. Can you add a check that to confirm that we output the right value?

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/CreditCardsAddressesStorage.kt [hunk 6]
```diff
@@ -192,7 +192,24 @@ data class CreditCard(
     val timeLastUsed: Long?,
     val timeLastModified: Long,
     val timesUsed: Long
-) : Parcelable
+) : Parcelable {
+    val obfuscatedCardNumber: String
```
Comment (gabrielluong) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/CreditCardsAddressesStorage.kt:196]: Perhaps add the same test for this too.

**Review by gabrielluong — APPROVED** [review]

**Pull request merged**